File: apps/HealthApp/shared/FitFusionCore/Sources/FitFusionCore/Intelligence/InsuranceCardOCR.swift
import Foundation
#if canImport(UIKit) && canImport(Vision)
import UIKit
import Vision

/// On-device insurance card OCR. Mirrors `NutritionLabelOCR.swift` —
/// `VNRecognizeTextRequest` plus a small regex set tuned for the
/// payer / member ID / group # / BIN / PCN / RxGrp fields most US
/// commercial cards print on the front face.
///
/// **Privacy:** image and OCR text are never uploaded. Parsed structured
/// fields land in the PHI Core Data store via `PHIStore.saveInsuranceCard`;
/// the raw OCR text lands in the Keychain via
/// `KeychainStore.Service.insurance` so it can be re-parsed without
/// re-snapping the card.
@MainActor
public final class InsuranceCardOCR {

    public static let shared = InsuranceCardOCR()
    private init() {}

    public struct Result: Sendable {
        public var payer: String?
        public var memberId: String?
        public var groupNumber: String?
        public var bin: String?
        public var pcn: String?
        public var rxGrp: String?
        public var rawText: String

        public init(payer: String? = nil, memberId: String? = nil,
                    groupNumber: String? = nil, bin: String? = nil,
                    pcn: String? = nil, rxGrp: String? = nil,
                    rawText: String = "") {
            self.payer = payer
            self.memberId = memberId
            self.groupNumber = groupNumber
            self.bin = bin
            self.pcn = pcn
            self.rxGrp = rxGrp
            self.rawText = rawText
        }
    }

    public func parse(image: UIImage) async -> Result {
        guard let cgImage = image.cgImage else { return .init() }
        return await withCheckedContinuation { cont in
            let request = VNRecognizeTextRequest { req, _ in
                let observations = (req.results as? [VNRecognizedTextObservation]) ?? []
                let lines = observations.compactMap { $0.topCandidates(1).first?.string }
                let joined = lines.joined(separator: "\n")
                cont.resume(returning: Self.extract(from: joined))
            }
            request.recognitionLevel = .accurate
            request.usesLanguageCorrection = true
            let handler = VNImageRequestHandler(cgImage: cgImage,
                                                orientation: .up, options: [:])
            DispatchQueue.global(qos: .userInitiated).async {
                do { try handler.perform([request]) }
                catch { cont.resume(returning: .init(rawText: "")) }
            }
        }
    }

    /// Pure-function extraction so it's unit-testable from JVM-style
    /// fixtures without needing a real `UIImage`.
    static func extract(from text: String) -> Result {
        var result = Result(rawText: text)

        // Member ID: look for "Member ID", "Member #", "Subscriber ID",
        // "ID #", or just "ID:". Captures alphanumeric token (8–20 chars).
        result.memberId = matchAfter(
            keys: ["member id", "member #", "member no", "subscriber id",
                   "subscriber #", "id #", "id:", "id no"],
            pattern: #"[A-Z0-9\-]{6,20}"#,
            in: text
        )
        // Group number
        result.groupNumber = matchAfter(
            keys: ["group #", "group no", "group:", "group number"],
            pattern: #"[A-Z0-9\-]{4,15}"#,
            in: text
        )
        // BIN — always 6 digits, often labelled "Rx BIN" / "BIN #".
        result.bin = matchAfter(
            keys: ["bin #", "bin:", "rx bin"],
            pattern: #"\d{6}"#,
            in: text
        )
        // PCN — alphanumeric, often labelled "PCN".
        result.pcn = matchAfter(
            keys: ["pcn:", "pcn #", "rx pcn"],
            pattern: #"[A-Z0-9]{2,15}"#,
            in: text
        )
        // RxGrp — alphanumeric, label "RxGrp" / "Rx Group".
        result.rxGrp = matchAfter(
            keys: ["rxgrp", "rx grp", "rx group"],
            pattern: #"[A-Z0-9\-]{2,15}"#,
            in: text
        )
        // Payer: first non-empty line that doesn't contain digits is a
        // decent heuristic for cards whose payer name is at the top.
        if let payer = text.split(separator: "\n")
            .map(String.init)
            .first(where: { line in
                let trimmed = line.trimmingCharacters(in: .whitespaces)
                return !trimmed.isEmpty &&
                       !trimmed.contains(where: \.isNumber) &&
                       trimmed.count > 2
            }) {
            result.payer = payer.trimmingCharacters(in: .whitespaces)
        }
        return result
    }

    /// Find the first occurrence of any key (case-insensitive) and return
    /// the regex match that follows it on the same line.
    private static func matchAfter(keys: [String], pattern: String, in text: String) -> String? {
        let lines = text.split(separator: "\n").map(String.init)
        let upper = text  // we case-insensitively match the keys via `.lowercased()` below
        _ = upper
        for line in lines {
            let lower = line.lowercased()
            guard let key = keys.first(where: { lower.contains($0) }) else { continue }
            let after = String(lower[lower.range(of: key)!.upperBound...])
            // Re-extract on the original-case substring of the line so the
            // returned token preserves original casing.
            let originalAfter = line.suffix(after.count)
            if let r = String(originalAfter).range(of: pattern, options: .regularExpression) {
                return String(originalAfter[r])
            }
        }
        return nil
    }
}
#endif

File: apps/HealthApp/android/app/src/main/java/com/myhealth/app/vision/InsuranceCardOcr.kt
package com.myhealth.app.vision

import android.graphics.Bitmap
import com.google.mlkit.vision.common.InputImage
import com.google.mlkit.vision.text.TextRecognition
import com.google.mlkit.vision.text.latin.TextRecognizerOptions
import javax.inject.Inject
import javax.inject.Singleton
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException
import kotlinx.coroutines.suspendCancellableCoroutine

/**
 * On-device insurance card OCR. Mirrors the iOS
 * [shared/.../Intelligence/InsuranceCardOCR.swift] regex set:
 * payer (top non-numeric line), member ID, group #, BIN, PCN, RxGrp.
 *
 * Image and OCR text never leave the device. Parsed structured fields go
 * into the SQLCipher PHI database via `InsuranceCardEntity`; raw OCR text
 * goes into `SecureTokenStore.setInsuranceRawText`.
 */
@Singleton
class InsuranceCardOcr @Inject constructor() {
    private val recognizer = TextRecognition.getClient(TextRecognizerOptions.DEFAULT_OPTIONS)

    data class Result(
        val payer: String? = null,
        val memberId: String? = null,
        val groupNumber: String? = null,
        val bin: String? = null,
        val pcn: String? = null,
        val rxGrp: String? = null,
        val rawText: String = "",
    )

    suspend fun read(bitmap: Bitmap, rotationDegrees: Int = 0): Result =
        suspendCancellableCoroutine { cont ->
            val image = InputImage.fromBitmap(bitmap, rotationDegrees)
            recognizer.process(image)
                .addOnSuccessListener { v -> cont.resume(extract(v.text)) }
                .addOnFailureListener { e -> cont.resumeWithException(e) }
        }

    fun extract(text: String): Result {
        val payer = text.lineSequence()
            .map { it.trim() }
            .firstOrNull { it.isNotEmpty() && it.length > 2 && it.none(Char::isDigit) }
        val memberId = matchAfter(
            keys = listOf("member id", "member #", "member no", "subscriber id",
                "subscriber #", "id #", "id:", "id no"),
            pattern = """[A-Z0-9\-]{6,20}""",
            text = text
        )
        val group = matchAfter(
            keys = listOf("group #", "group no", "group:", "group number"),
            pattern = """[A-Z0-9\-]{4,15}""",
            text = text
        )
        val bin = matchAfter(
            keys = listOf("bin #", "bin:", "rx bin"),
            pattern = """\d{6}""",
            text = text
        )
        val pcn = matchAfter(
            keys = listOf("pcn:", "pcn #", "rx pcn"),
            pattern = """[A-Z0-9]{2,15}""",
            text = text
        )
        val rxGrp = matchAfter(
            keys = listOf("rxgrp", "rx grp", "rx group"),
            pattern = """[A-Z0-9\-]{2,15}""",
            text = text
        )
        return Result(
            payer = payer, memberId = memberId, groupNumber = group,
            bin = bin, pcn = pcn, rxGrp = rxGrp, rawText = text
        )
    }

    private fun matchAfter(keys: List<String>, pattern: String, text: String): String? {
        val regex = Regex(pattern, RegexOption.IGNORE_CASE)
        for (line in text.split("\n")) {
            val lower = line.lowercase()
            val key = keys.firstOrNull { lower.contains(it) } ?: continue
            val tail = line.substring(lower.indexOf(key) + key.length)
            return regex.find(tail)?.value
        }
        return null
    }
}
